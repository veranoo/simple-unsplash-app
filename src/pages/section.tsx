import React, { memo, useCallback, useEffect, useReducer, useRef } from 'react';
import { Link, RouteComponentProps } from 'react-router-dom';
import InfiniteScroll from 'react-infinite-scroller';

import styled from 'styled-components';
import { PhotoImage } from '../components/photo-image';
import { Layout } from '../components/layout';
import Container from '../components/container';
import {
  sectionReducer,
  SET_ERROR,
  SET_LOAD_MORE,
  SET_LOAD_MORE_ERROR,
  SET_NOT_LOAD_MORE,
  SET_PHOTOS
} from '../reducers/section-reducer';
import { useUnsplashApi } from '../hooks/use-unsplash-api';

const ImagesWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
`;

const ImageWrapper = styled.div`
  width: 22%;
  margin: 10px 0;
  position: relative;
`;

const SelectWrapper = styled.div`
  display: flex;
  justify-content: center;
  height: 50px;
  align-items: center;
`;

interface PhotoItemProps {
  id: number;
  urls: {
    small: string;
  };
  alt?: string;
}

const PhotoItem = memo<PhotoItemProps>(({ id, urls, alt }) => {
  return (
    <ImageWrapper key={id} className='image'>
      <Link to={`/photo/${id}`}>
        <PhotoImage id={id} src={urls.small} alt={alt} />
      </Link>
    </ImageWrapper>
  );
});

const INITIAL_PAGE = 1;
const PHOTOS_PER_PAGE = 10;
const INITIAL_ORDER_BY = 'latest';

export const Section: React.FC<RouteComponentProps<{ id: string }>> = props => {
  const unsplashApi = useUnsplashApi();
  const params = useRef({
    page: INITIAL_PAGE,
    perPage: PHOTOS_PER_PAGE,
    orderBy: INITIAL_ORDER_BY
  });

  const ref = useRef();

  const [state, dispatch] = useReducer(sectionReducer, {
    photos: [],
    hasMore: false,
    error: false
  });

  const fetchCollections = useCallback(() => {
    unsplashApi
      .getCollectionPhotos(
        Number(props.match.params.id),
        params.current.page,
        params.current.perPage,
        params.current.orderBy
      )
      .then(response => {
        dispatch({ type: SET_PHOTOS, payload: { photos: response } });
      })
      .catch(() => {
        if (!ref.current) {
          return;
        }
        dispatch({ type: SET_ERROR });
      });
  }, []);

  useEffect(() => {
    fetchCollections();

    return () => {
      unsplashApi.abort();
    }
  }, []);

  const loadMore = useCallback(() => {
    params.current.page += 1;

    unsplashApi
      .getCollectionPhotos(
        Number(props.match.params.id),
        params.current.page,
        params.current.perPage,
        params.current.orderBy
      )
      .then(response => {
        if (!response.length) {
          dispatch({ type: SET_NOT_LOAD_MORE });
          return;
        }

        dispatch({ type: SET_LOAD_MORE, payload: { photos: response } });
      })
      .catch(() => {
        if (!ref.current) {
          return;
        }
        dispatch({ type: SET_LOAD_MORE_ERROR });
      });
  }, []);

  const handleChangeOrder = useCallback(evt => {
    params.current.orderBy = evt.target.value;
    fetchCollections();
  }, []);

  const items = (
    <ImagesWrapper>
      {state.photos.map(item => (
        <PhotoItem key={item.id} id={item.id} urls={item.urls} />
      ))}
    </ImagesWrapper>
  );

  return (
    <Layout>
      <SelectWrapper ref={ref}>
        <select onChange={handleChangeOrder}>
          <option value='latest'>Najnowsze</option>
          <option value='popular'>Popularne</option>
        </select>
      </SelectWrapper>
      <Container>
        <InfiniteScroll
          loadMore={loadMore}
          hasMore={state.hasMore}
          loader={<div key='loading'>Loading..</div>}
        >
          {items}
        </InfiniteScroll>
        {state.error && <div key='error'>Wystąpił błąd</div>}
      </Container>
    </Layout>
  );
};
