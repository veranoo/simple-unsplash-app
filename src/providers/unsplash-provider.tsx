import React, { useMemo } from 'react';
import Unsplash from 'unsplash-js';

interface UnsplashProviderInterface {
  listCollections(): Promise<any[]>;
  getPhoto(photoId: string): Promise<any>;
  getCollectionPhotos(
    collectionId: number,
    page: number,
    perPage: number,
    orderBy: string
  ): Promise<any>;
}

export const UnsplashContext = React.createContext<UnsplashProviderInterface>(null);

export const UnsplashProvider: React.FC = ({ children }) => {
  const unsplash = useMemo(() => {
    const unsplashInstance = new Unsplash({
      accessKey: process.env.ACCES_KEY
    });

    return {
      listCollections: () => {
        return unsplashInstance.collections
          .listCollections()
          .then(res => res.json());
      },
      getPhoto: photoId => {
        return unsplashInstance.photos
          .getPhoto(photoId)
          .then(res => res.json());
      },
      getCollectionPhotos: (...args) => {
        return unsplashInstance.collections
          .getCollectionPhotos(...args)
          .then(res => res.json());
      }
    };
  }, []);

  return (
    <UnsplashContext.Provider value={unsplash}>
      {children}
    </UnsplashContext.Provider>
  );
};
