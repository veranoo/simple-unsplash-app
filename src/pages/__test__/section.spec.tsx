import { render, wait, fireEvent } from '@testing-library/react';
import { UnsplashContext } from '../../providers/unsplash-provider';
import React from 'react';
import collections from '../../../mocks/collections.json';
import photos from '../../../mocks/photos.json';
import { Section } from '../section';
import { BrowserRouter } from 'react-router-dom';
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';

test('should render Section page when success', async () => {
  const mock = {
    listCollections: () => Promise.resolve(collections),
    getPhoto: () => Promise.resolve(null),
    getCollectionPhotos: () => Promise.resolve(photos)
  };
  const MockProvider = ({ children }) => (
    <UnsplashContext.Provider value={mock as any}>{children}</UnsplashContext.Provider>
  );

  jest.spyOn(mock, 'getCollectionPhotos').mockImplementationOnce(() => {
    return Promise.resolve(photos);
  });

  mockAllIsIntersecting(true);

  const { baseElement, getAllByRole, debug,  } = render(
    <BrowserRouter>
      <MockProvider>
        <Section
          history={null as any}
          location={null as any}
          match={{ params: { id: 0 } } as any}
        />
      </MockProvider>
    </BrowserRouter>
  );

  await wait();
  mockAllIsIntersecting(true);
  await wait();
  expect(mock.getCollectionPhotos).toBeCalled();
  debug();
});
