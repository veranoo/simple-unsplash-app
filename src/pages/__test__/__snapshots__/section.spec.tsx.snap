// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Section page when success 1`] = `
<div>
  <div
    class="sc-bxivhb glCqXO"
  >
    <nav
      class="sc-ifAKCX bHTZJD"
    >
      <a
        href="/"
      >
        <img
          alt=""
          src="http://assets.stickpng.com/thumbs/5cb4839d5f1b6d3fbadece7c.png"
          width="80px"
        />
      </a>
    </nav>
    <div
      class="sc-htoDjs bIJLve"
    >
      <select>
        <option
          value="latest"
        >
          Najnowsze
        </option>
        <option
          value="popular"
        >
          Popularne
        </option>
      </select>
    </div>
    <div
      class="sc-EHOje jtMLyu"
    >
      <div>
        <div
          class="sc-bZQynM haTaOE"
        />
      </div>
    </div>
  </div>
</div>
`;
